src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,846 @@
+
+            this.userNameBytes = null != SQLServerDriverStringProperty.USER.getDefaultValue() ? unicode(userName)
+                                                                                              : null;
+            this.upperUserName = userName.toUpperCase();
+
+            this.passwordHash = null != password ? MD4(unicode(password)) : null;
+
+            this.workstationBytes = null != workstation ? unicode(workstation.toUpperCase()) : null;
+
+            try {
+                mac = Mac.getInstance("HmacMD5");
+            } catch (NoSuchAlgorithmException e) {
+                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmHmacMD5Error"));
+                Object[] msgArgs = {domainName, e.getMessage()};
+                throw new SQLServerException(form.format(msgArgs), e);
+            }
+        }
+    };
+
+    // Handle to NTLM context
+    private NTLMContext context = null;
+
+    /**
+     * Creates an instance of the NTLM authentication
+     *
+     * @param domainName
+     * @param userName
+     * @param password
+     * @param workstation
+     * @throws SQLServerException
+     */
+    NTLMAuthentication(final String domainName, final String userName, final String password,
+            final String workstation) throws SQLServerException {
+        if (null == context) {
+            this.context = new NTLMContext(domainName, userName, password, workstation);
+        }
+    }
+
+    /**
+     * Generates the NTLM client context
+     *
+     * @param inToken
+     *        SSPI input blob
+     * @param done
+     *        indicates processing is done
+     * @return NTLM client context
+     * @throws SQLServerException
+     */
+    @Override
+    byte[] generateClientContext(final byte[] inToken, final boolean[] done) throws SQLServerException {
+        return initializeSecurityContext(inToken, done);
+    }
+
+    /**
+     * Releases the NTLM client context
+     *
+     * @throws SQLServerException
+     */
+    @Override
+    int releaseClientContext() throws SQLServerException {
+        context = null;
+        return 0;
+    }
+
+    /**
+     * Parses the Type 2 NTLM Challenge message from server
+     *
+     * Section 2.2.1.2 CHALLENGE_MESSAGE
+     *
+     * @param inToken
+     *        SSPI input blob
+     * @throws SQLServerException
+     */
+    private void parseNtlmChallenge(final byte[] inToken) throws SQLServerException {
+        // get token buffer
+        ByteBuffer token = ByteBuffer.wrap(inToken).order(ByteOrder.LITTLE_ENDIAN);
+
+        // verify signature
+        byte[] signature = new byte[NTLM_HEADER_SIGNATURE.length];
+        token.get(signature);
+        if (!Arrays.equals(signature, NTLM_HEADER_SIGNATURE)) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmSignatureError"));
+            Object[] msgArgs = {new String(signature, java.nio.charset.StandardCharsets.UTF_16LE)};

```

Comment:
Does it make sense to convert NTLM signature to String? I think it should rather be a hex representation.